File: Leetcode/HumbleNumbers/76_Minimum_Window_Substring/76_Minimum_Window_Substring.py

from collections import defaultdict
class Solution:
     def minWindow(self, s: str, t: str) -> str:
        target=defaultdict(int)
        for c in t:
            target[c]+=1
        min_len=float('inf')
        counter=len(t)
        left=0
        right=0
        result=""
        while right<len(s):
            if target[s[right]]>0:
                counter-=1
            target[s[right]]-=1
            right+=1
            while counter==0:
                if min_len>right-left:
                    min_len=right-left
                    result=s[left:right]
                if target[s[left]]==0:
                    counter+=1
                target[s[left]]+=1
                left+=1     
        return result         
def main():
        s = "TMFRAFDOBECODEBANC"
        t = "ABC"
        
        s1="a"
        t1="aa"
        m=Solution()
        print(m.minWindow(s1,t1))
        
if __name__=='__main__':
    main()
        
                        
                    
                    
                    
            
        
            
        

File: Leetcode/HumbleNumbers/76_Minimum_Window_Substring/C++/76_Minimum_Window_Substring.cpp
#include<iostream>
#include<unordered_map>
using namespace std;
class Solution {
public:
    std::string minWindow(std::string s, std::string t) {
        std::string result;
        std::unordered_map<char,int>words,windows;
        int left=0;
        
        for(auto& word:t)
        {
            words[word]+=1;
        }
        int count=0;
        for( int i=0;i<s.size();i++)
        { 
            
            windows[s[i]]++;
            if(windows[s[i]]==words[s[i]])
            {
                count+=1;
            }
            while(windows[s[left]]>words[s[left]])
            {
                
                windows[s[left++]]--;
            }
            if(count==t.size())
            {
                auto flag=result.empty()||i-left+1<result.size();
                if(result.empty()||i-left+1<result.size())
                {
                    result=s.substr(left,i-left+1);
                    
                }
            }
        }
        return result;
    }
    



};
int main(int argc, char const *argv[])
{
    /* code */
    std::string m="ADOBECODEBANC";
    std::string n="ABC";
    Solution s;
    std::string &&t=s.minWindow(m,n);
    
    std::cerr<<"result= "<<t<<std::endl;
   
    return 0;
}

